Environment_Engine/Query/Panels.cs

```diff
@@ -290,5 +290,121 @@ public static List<Panel> PanelsFromPhysical(this List<BH.oM.Physical.Elements.I
 
             return panels;
         }
+
+        [Description("Returns the floor panels of a space represented by Environment Panels")]
+        [Input("panelsAsSpace", "A collection of Environment Panels that represent a closed space")]
+        [Output("floorPanels", "BHoM Environment panel representing the floor of the space")]
+        public static List<Panel> FloorPanels(this List<Panel> panelsAsSpace)
+        {
+            //Find the panel(s) that are at the lowest point of the space...
+
+            double minZ = 1e10;
+            foreach (Panel panel in panelsAsSpace)
+            {
+                if (panel.MinimumLevel() == panel.MaximumLevel())
+                    minZ = Math.Min(minZ, panel.MinimumLevel());
+            }
+
+            List<Panel> floorPanels = panelsAsSpace.Where(x => x.MinimumLevel() == minZ && x.MaximumLevel() == minZ).ToList();
+
+            foreach (Panel panel in floorPanels)
+            {
+                if (panel.ConnectedSpaces.Where(x => x != "-1").ToList().Count == 1)
+                    panel.Type = PanelType.SlabOnGrade;
+                else if (panel.ConnectedSpaces.Where(x => x != "-1").ToList().Count == 2)
+                    panel.Type = PanelType.FloorInternal;
+
+                if (floorPanels.Count == 0)
+                {
+                    BH.Engine.Reflection.Compute.RecordWarning("Could not find roof panel");
+                    return null;
+                }
+            }
+
+            return floorPanels;
+        }
+
+        [Description("Returns the roof panels of a space represented by Environment Panels")]
+        [Input("panelsAsSpace", "A collection of Environment Panels that represent a closed space")]
+        [Output("roofPanels", "BHoM Environment panel representing the roof of the space")]
+        public static List<Panel> RoofPanels(this List<Panel> panelsAsSpace)
+        {
+            //Find the panel(s) that are at the highest point of the space...
+
+            double minZ = 1e10;
+            foreach (Panel panel in panelsAsSpace)
+            {
+                if (panel.MinimumLevel() == panel.MaximumLevel())
+                    minZ = Math.Min(minZ, panel.MinimumLevel());
+            }
+
+            List<Panel> roofPanels = panelsAsSpace.Where(x => ((x.MinimumLevel() != minZ && x.MaximumLevel() != minZ) && Math.Round(x.Tilt()) != 90) && x.ConnectedSpaces.ToList().Count == 1).ToList();
+
+            foreach (Panel panel in panelsAsSpace)
```

Comment:
I think this `for` loop is incorrect, this should be on `roofPanels` rather than all of the space panels?